File: JAVA/27_ShortestDistanceBacktracking/BOJ_2568_전깃줄2.java
import java.io.IOException;
import java.io.BufferedReader;
import java.io.InputStreamReader;
import java.util.StringTokenizer;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Stack;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        int n = Integer.parseInt(br.readLine());
        int[] poleA = new int[n + 1];
        int[] arr = new int[500001];
        int[] idxArr = new int[n + 1];
        for (int i = 1; i <= n; i++) {
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            poleA[i] = Integer.parseInt(st.nextToken());
            arr[poleA[i]] = Integer.parseInt(st.nextToken());
        }
        Arrays.sort(poleA);
        ArrayList<Integer> list = new ArrayList<>();
        list.add(Integer.MIN_VALUE);
        for (int i = 1; i <= n; i++) {
            if (arr[poleA[i]] > list.get(list.size() - 1)) {
                idxArr[i] = list.size();
                list.add(arr[poleA[i]]);
            } else {
                int left = 1, right = list.size() - 1, idx = 1;
                while (left <= right) {
                    int mid = (left + right) / 2;
                    if (list.get(mid) >= arr[poleA[i]]) {
                        idx = mid;
                        right = mid - 1;
                    } else {
                        left = mid + 1;
                    }
                }
                list.set(idx, arr[poleA[i]]);
                idxArr[i] = idx;
            }
        }
        int max = list.size() - 1;
        StringBuilder sb = new StringBuilder();
        sb.append(n - max).append("\n");
        Stack<Integer> stack = new Stack<>();
        for (int i = n; i > 0; i--) {
            if (idxArr[i] == max) max--;
            else stack.push(poleA[i]);
        }
        while (!stack.isEmpty()) {
            sb.append(stack.pop()).append("\n");
        }
        System.out.print(sb);
    }
}

File: C/27_ShortestDistanceBacktracking/BOJ_2568_전깃줄2.c
#include <stdio.h>
#include <stdlib.h>
int arr[500001], stackIdx = -1, * stack;
void push(int value) {
	stack[++stackIdx] = value;
}
int is_empty() {
	if (stackIdx < 0) return 1;
	else return 0;
}
int pop() {
	if (is_empty()) return -1;
	else return stack[stackIdx--];
}
int compare(const void* a, const void* b) {
	int o1 = *(int*)a;
	int o2 = *(int*)b;
	if (o1 > o2) return 1;
	else if (o1 < o2) return -1;
	else return 0;
}
main() {
	int n, top = -1;
	scanf("%d", &n);
	int* poleA = (int*)malloc(sizeof(int) * (n + 1));
	int* idxArr = (int*)malloc(sizeof(int) * (n + 1));
	int* list = (int*)malloc(sizeof(int) * (n + 1));
	for (int i = 1; i <= n; i++) {
		scanf("%d", &poleA[i]);
		scanf("%d", &arr[poleA[i]]);
	}
	qsort(poleA, n + 1, sizeof(int), compare);
	list[++top] = 0;
	for (int i = 1; i <= n; i++) {
		if (arr[poleA[i]] > list[top]) {
			idxArr[i] = ++top;
			list[top] = arr[poleA[i]];
		}
		else {
			int left = 1, right = top, idx = 1;
			while (left <= right) {
				int mid = (right + left) / 2;
				if (list[mid] >= arr[poleA[i]]) {
					idx = mid;
					right = mid - 1;
				}
				else {
					left = mid + 1;
				}
			}
			list[idx] = arr[poleA[i]];
			idxArr[i] = idx;
		}
	}
	printf("%d\n", n - top);
	stack = (int*)malloc(sizeof(int) * (n - top));
	for (int i = n; i > 0; i--) {
		if (idxArr[i] == top) top--;
		else push(poleA[i]);
	}
	while (!is_empty()) {
		printf("%d\n", pop());
	}
}